File: facet_mcp/protocol/transport.py
"""
MCP Transport Layer

Provides transport mechanisms for MCP communication:
- JSON-RPC over WebSockets
- HTTP/JSON endpoints
- gRPC support (future)

Handles connection management, message serialization, and error handling.
"""

import asyncio
import json
import logging
from typing import Callable, Optional
import websockets
from websockets.exceptions import ConnectionClosedError

from .messages import MCPMessage

logger = logging.getLogger(__name__)


class MCPTransport:
    """
    MCP Transport Layer

    Handles the low-level communication details for MCP protocol.
    Supports WebSocket connections for real-time communication.
    """

    def __init__(self):
        self.server: Optional[asyncio.AbstractServer] = None
        self.message_handler: Optional[Callable[[MCPMessage], MCPMessage]] = None
        self.connections = set()

    async def start_server(
        self,
        host: str = "localhost",
        port: int = 3000,
        message_handler: Optional[Callable[[MCPMessage], MCPMessage]] = None
    ):
        """
        Start MCP server on specified host and port.

        Args:
            host: Server host address
            port: Server port
            message_handler: Function to handle incoming MCP messages
        """
        self.message_handler = message_handler

        # Start WebSocket server
        self.server = await websockets.serve(
            self._handle_connection,
            host,
            port,
            ping_interval=30,  # Keep connections alive
            ping_timeout=10,
            close_timeout=5
        )

        logger.info(f"MCP server started on {host}:{port}")
        logger.info("Available endpoints:")
        logger.info(f"  WebSocket: ws://{host}:{port}")
        logger.info(f"  HTTP: http://{host}:{port}/health")

        # Keep server running
        await self.server.wait_closed()

    async def stop_server(self):
        """Stop the MCP server gracefully"""
        if self.server:
            self.server.close()
            await self.server.wait_closed()
            logger.info("MCP server stopped")

    async def _handle_connection(self, websocket):
        """
        Handle individual WebSocket connection.

        Manages the lifecycle of a single client connection,
        including message handling and connection cleanup.
        """
        connection_id = id(websocket)
        self.connections.add(websocket)

        try:
            logger.info(f"New MCP connection established: {connection_id}")

            async for message_raw in websocket:
                try:
                    # Parse incoming message
                    message_data = json.loads(message_raw)
                    message = MCPMessage.from_dict(message_data)

                    logger.debug(f"Received message: {message.type}")

                    # Handle message
                    if self.message_handler:
                        response = await self.message_handler(message)

                        # Send response
                        response_data = response.to_dict()
                        await websocket.send(json.dumps(response_data))
                        logger.debug(f"Sent response: {response.type}")
                    else:
                        # No handler - send error
                        error_response = MCPMessage(
                            type="error",
                            data={"error": "No message handler configured"}
                        )
                        await websocket.send(json.dumps(error_response.to_dict()))

                except json.JSONDecodeError as e:
                    logger.error(f"Invalid JSON received: {e}")
                    error_response = MCPMessage(
                        type="error",
                        data={"error": "Invalid JSON format"}
                    )
                    await websocket.send(json.dumps(error_response.to_dict()))

                except Exception as e:
                    logger.error(f"Error handling message: {e}")
                    error_response = MCPMessage(
                        type="error",
                        data={"error": f"Internal server error: {str(e)}"}
                    )
                    await websocket.send(json.dumps(error_response.to_dict()))

        except ConnectionClosedError:
            logger.info(f"MCP connection closed: {connection_id}")
        except Exception as e:
            logger.error(f"Connection error: {e}")
        finally:
            # Cleanup connection
            self.connections.discard(websocket)
            logger.info(f"Connection cleanup completed: {connection_id}")


class MCPClient:
    """
    MCP Client for connecting to MCP servers.

    Allows AI agents and other tools to connect to and use MCP servers.
    """

    def __init__(self):
        self.websocket: Optional[websockets.WebSocketServerProtocol] = None

    async def connect(self, uri: str):
        """
        Connect to MCP server.

        Args:
            uri: WebSocket URI (e.g., "ws://localhost:3000")
        """
        try:
            self.websocket = await websockets.connect(uri)
            logger.info(f"Connected to MCP server: {uri}")
        except Exception as e:
            logger.error(f"Failed to connect to MCP server: {e}")
            raise

    async def disconnect(self):
        """Disconnect from MCP server"""
        if self.websocket:
            await self.websocket.close()
            logger.info("Disconnected from MCP server")

    async def send_message(self, message: MCPMessage) -> MCPMessage:
        """
        Send message to MCP server and wait for response.

        Args:
            message: MCP message to send

        Returns:
            Response message from server
        """
        if not self.websocket:
            raise ConnectionError("Not connected to MCP server")

        try:
            # Send message
            message_data = message.to_dict()
            await self.websocket.send(json.dumps(message_data))

            # Wait for response
            response_raw = await self.websocket.recv()
            response_data = json.loads(response_raw)
            response = MCPMessage.from_dict(response_data)

            return response

        except Exception as e:
            logger.error(f"Error sending message: {e}")
            raise

    async def call_tool(
        self,
        tool_name: str,
        parameters: dict,
        timeout: float = 30.0
    ) -> dict:
        """
        Call a tool on the MCP server.

        Args:
            tool_name: Name of the tool to call
            parameters: Tool parameters
            timeout: Timeout in seconds

        Returns:
            Tool execution result
        """
        from .messages import create_tool_call_message

        message = create_tool_call_message(tool_name, parameters)
        response = await asyncio.wait_for(
            self.send_message(message),
            timeout=timeout
        )

        if response.type == "tool_result":
            result_data = response.data
            if result_data.get("success", False):
                return result_data["result"]
            else:
                raise Exception(f"Tool execution failed: {result_data.get('error', 'Unknown error')}")
        else:
            raise Exception(f"Unexpected response type: {response.type}")

    async def list_tools(self) -> list:
        """
        Get list of available tools from MCP server.

        Returns:
            List of available tools
        """
        message = MCPMessage(type="list_tools", data={})
        response = await self.send_message(message)

        if response.type == "tools_list":
            return response.data.get("tools", [])
        else:
            raise Exception(f"Unexpected response type: {response.type}")


# Health check endpoint for HTTP requests
async def health_check(request):
    """Simple health check endpoint"""
    return {
        "status": "healthy",
        "service": "FACET MCP Server",
        "version": "0.1.0"
    }

File: npm_package/src/protocol/transport.ts
import WebSocket from 'ws';
import { EventEmitter } from 'events';
import { MCPMessage, ToolCall, ToolResult, ToolError } from '../types';
import MessageFactory from './messages';

export interface TransportConfig {
  host: string;
  port: number;
  timeout: number;
  maxReconnectAttempts: number;
  reconnectInterval: number;
}

export class MCPTransport extends EventEmitter {
  private ws: WebSocket | null = null;
  private config: TransportConfig;
  private reconnectAttempts = 0;
  private reconnectTimer?: NodeJS.Timeout;
  private pendingRequests = new Map<string, {
    resolve: (value: any) => void;
    reject: (error: any) => void;
    timeout: NodeJS.Timeout;
  }>();

  constructor(config: Partial<TransportConfig> = {}) {
    super();
    this.config = {
      host: 'localhost',
      port: 3000,
      timeout: 30000,
      maxReconnectAttempts: 5,
      reconnectInterval: 1000,
      ...config
    };
  }

  async connect(): Promise<void> {
    const url = `ws://${this.config.host}:${this.config.port}`;

    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(url);

        this.ws.on('open', () => {
          console.log(`Connected to MCP server at ${url}`);
          this.reconnectAttempts = 0;
          this.emit('connected');
          resolve();
        });

        this.ws.on('message', (data: WebSocket.Data) => {
          this.handleMessage(data);
        });

        this.ws.on('error', (error) => {
          console.error('WebSocket error:', error);
          this.emit('error', error);
          reject(error);
        });

        this.ws.on('close', (code, reason) => {
          console.log(`Disconnected from MCP server (code: ${code}, reason: ${reason})`);
          this.emit('disconnected', code, reason);
          this.handleReconnection();
        });

      } catch (error) {
        reject(error);
      }
    });
  }

  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }

    // Clear pending requests
    for (const [id, request] of this.pendingRequests) {
      clearTimeout(request.timeout);
      request.reject(new Error('Connection closed'));
    }
    this.pendingRequests.clear();
  }

  async sendToolCall(tool: string, params: Record<string, any>): Promise<ToolResult> {
    const message = MessageFactory.createToolCall(tool, params);

    return new Promise((resolve, reject) => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        reject(new Error('WebSocket not connected'));
        return;
      }

      const timeout = setTimeout(() => {
        this.pendingRequests.delete(message.id);
        reject(new Error(`Request timeout after ${this.config.timeout}ms`));
      }, this.config.timeout);

      this.pendingRequests.set(message.id, { resolve, reject, timeout });

      this.ws.send(MessageFactory.serializeMessage(message));
    });
  }

  private handleMessage(data: WebSocket.Data): void {
    const message = MessageFactory.parseMessage(data.toString());

    if (!message) {
      console.error('Received invalid message:', data);
      return;
    }

    this.emit('message', message);

    // Handle pending requests
    if (message.type === 'tool_result' || message.type === 'error') {
      const request = this.pendingRequests.get(message.id);
      if (request) {
        clearTimeout(request.timeout);
        this.pendingRequests.delete(message.id);

        if (message.type === 'tool_result') {
          const result = message as ToolResult;
          if (result.success) {
            request.resolve(result.result);
          } else {
            request.reject(new Error(`Tool execution failed: ${result.result}`));
          }
        } else {
          const error = message as ToolError;
          request.reject(new Error(`${error.code}: ${error.error}`));
        }
      }
    }
  }

  private handleReconnection(): void {
    if (this.reconnectAttempts >= this.config.maxReconnectAttempts) {
      this.emit('maxReconnectAttemptsReached');
      return;
    }

    this.reconnectAttempts++;
    console.log(`Attempting to reconnect (${this.reconnectAttempts}/${this.config.maxReconnectAttempts})...`);

    this.reconnectTimer = setTimeout(() => {
      this.connect().catch(error => {
        console.error('Reconnection failed:', error);
      });
    }, this.config.reconnectInterval * this.reconnectAttempts);
  }

  get isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }

  get connectionState(): string {
    if (!this.ws) return 'disconnected';

    switch (this.ws.readyState) {
      case WebSocket.CONNECTING: return 'connecting';
      case WebSocket.OPEN: return 'connected';
      case WebSocket.CLOSING: return 'closing';
      case WebSocket.CLOSED: return 'disconnected';
      default: return 'unknown';
    }
  }
}

export default MCPTransport;
